### ISSUE TITLE
Using multiple components in different modules causing "Type X is part of the declarations of 2 modules" error

### ISSUE THREAD

**Issue by pleerock** [angular/angular#10646]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

Right now you cannot import two directives of the same module 

**Expected/desired behavior**

You can use same directives/components in multiple modules without errors.

**What is the motivation / use case for changing the behavior?**

Concept of the shared module with its shared resources is good for components that are used everywhere. But if you have a complex application (ours contain more then 50 modules and 1500 components, services and pipes all together) multiple components can be reused different way. For example QuestionListPage uses UserActiveListPanel from User module. Also PhotoListPage uses same UserActiveListPanel, and VideoListPage uses UserActiveListPanel. How do I reuse this panel? Best way is see is to provide it in `directives` of the component as it was before rc.5. But right now I can't do it. But I can create lets say UserShareModule and put it there... its good, but if I put everything that is used by other modules from user module, i'll LOT of things there unnecessary in most of times. Then what to do? Create UserActiveListPanelShareModule? It will create lot of such trash files. So what to do? Maybe we simply can provide same directives multiple times without having this annoying error `Type X is part of the declarations of 2 modules`?

If you tell me that its by design, then my answer is: in my opinion its a bad design.

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.5
 
* **Browser:** [all ] 
 
* **Language:** [all]

**Comment by brandonroberts** [comment]
If you want to use a component across two modules, you'll need to create a "shared" module and add that component the shared module's `exports`. Then you import that shared module into your other modules.

**Comment by pleerock** [comment]
yeah I did it and it works. The problem is that it is very very inconvenient for multiple directives/components that are used only few times by other components.

**Comment by brandonroberts** [comment]
Once you have your shared module setup it should be easier to add
components to multiple modules. In the long run it helps you to organize
your code into modules with a standard convention.

**Comment by pleerock** [comment]
No it really harms organization in a long run.

Lets bring an example. I have two modules. `QuestionModule`, `PostModule` and `UserModule`. `QuestionModule` and `PostModule` need `UserListPanelComponent` from a `UserModule`. How to handle such situation? 

**option1.** Create a `UserSharedModule` and add there `UserListPanelComponent`. But it looks like we gonna put to `UserSharedModule` all other components too (lets say other 20 components out of 50 `UserModule` has), because other modules (30) also uses different components of the `UserModule`. QuestionModule (or any other that uses UserModule's components) loads lot of stuff that he don't really need. It loads all 20 components when it needs only one. So it seems to be a bad option.

**option2.** Create `UserListPanelSharedModule`. that shares only one component -`UserListPanelComponent`. Now I can load only one file and not worry that it loads all 20 components. But wait... its annoying and simply stupid to create one another separate file for each component you want to share. So, it definitely bad option.

Any good options here?

**Comment by jusefb** [comment]
This is indeed very confusing. I have moved some functionality into a module and now I am getting this error: "NgSelectOption is part of the declarations of 2 modules: InternalFormsSharedModule and SpFormsModule!" SpFormsModule is my module but I am not importing InternalFormsSharedModule anywhere. Any idea  why would these two modules clash?

**Comment by brandonroberts** [comment]
@jusefb this usually means you've added the `NgSelectOption` to your `declarations` instead of importing the appropriate forms module. The rule is if you created the directive, you need to declare it. If you didn't create it and it comes with a module, it doesn't go in your `declarations`. Instead you add that module to your `imports`.

For your previous comment, if you are concerned with modules sharing components that other modules don't need, then you organize further. I used the term "shared" module loosely because there doesn't have to be just one. You can create a set of small modules to import into other modules to share components.

**Comment by DaSchTour** [comment]
If you want to use the same Component you can maybe try to cheat angular by using extends.
Try something like this.

```typescript
@Component({
    selector: "selector",
    templateUrl: "/component.html"
})
class BaseComponent {}

export class AComponent extends BaseComponent {}
export class BComponent extends BaseComponent {}
```

No you have two components that are basically the same.
To avoid conflicts you might add different selectors to the components.

**Comment by vicb** [comment]
Your question sounds like a support request.

Please use the issue tracker only for bugs and feature requests.

Use [gitter](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem) and [StackOverflow](http://stackoverflow.com/questions/tagged/angular) for support request.

**Comment by pleerock** [comment]
where do you see a question here?

**Comment by PunkSage** [comment]
@vicb I think that you need to read more carefully before you close an issue.

**Comment by ootwch** [comment]
Same issue here: "Type NgSelectOption is part of the declarations of 2 modules".  Fulltext search through my code has not found any occurence of 'NgSelectOption', so I am pretty sure that this is caused by a third party module (?) or library... but which one?

=> Please add "better error message" to this feature request, I think it might make the migration to post rc.5 with ngmodules a lot less painful

Note: The solutions presented in https://github.com/angular/angular/issues/10784 did not help either...

**Comment by SnareChops** [comment]
Agreed, this is a legitimate feature request and not a question requesting support. This should be reopened for discussion.

**Comment by brandonroberts** [comment]
@PunkSage @ootwch @SnareChops can someone reproduce this in a plunker? It would help in getting the issue reopened. If its a third party module that's causing it, my guess is that two modules  have the `NgSelectOption` in their `declarations` instead of importing the appropriate module.

**Comment by kylecordes** [comment]
@brandonroberts and others - I think this error occurs when a component (either user-written or built-in) has been accidentally, implicitly, *transitively* included in more than one module - which can easily happen even without the now-deprecated "directives" list. Maybe that tip will make it easier to make a repro plunker. Still, I suspect it will turn out to be a "feature, not a bug".

**Comment by brandonroberts** [comment]
@kylecordes Right, but if you've upgraded to RC5 and you haven't removed all your `directives` from your components, and one of those components is still including one of the old FORM_DIRECTIVES arrays, I could see why it would happen.

**Comment by ootwch** [comment]
@brandonroberts Thanks, that has bee helpful. I guess it could also be `FormControl` or `ControlValueAccessor`? 

New full-text search on `@angular/forms` found that some of the imported components included `@angular/forms`. Removing the components where those are used seems to make the error message disappear.

Note that I was mentioning the error message above, seems in one case it gave me the name of the offending module... in another case it mentioned  `InternalFormsSharedModule`which was less helpful. However, this was with angular2-material, which offers a module-based version I should update to...

Conclusion: The update from RC4 to RC5 requires to update (all?) third party libraries in use, or at least the ones designed for ng2? Seems the whole thing is less backward compatible than I had been hoping from the documentation ;-), but it also seems solvable now...

Would it have been helpful to have `import { ... } from "@angular/forms";` provide an error or warning? That might be a feature request for a future version... :-)

**Comment by kylecordes** [comment]
@ootwch With the new module system, it will be most typical for third-party Angular based libraries to present themselves to your application as a module. So yes, they will need to be updated to embrace RC5+. However, in the meantime there is a fairly easy workaround. You can simply add a single file in your application which defines a module which includes the contents of the older (RC4 or less) angular based library you were using. Then depend on this module from each of your own modules. That should get you past the trouble between now and when those libraries are updated.

**Comment by lacma2011** [comment]
I am having problems with ng2-select. I can't drop its directive into multiple modules now in rc5. It was a good library for making nice multi-select dropdowns but now it's ridiculous that it can't be used twice in the application.

**Comment by jcmordan** [comment]
Hi, I have the same problem here. I try moving the common components to the AppModule. but then i got and error 
 'Can't bind to 'option' since it isn't a known property of 'panel-deatail'.

<panel-deatail option="data"></panel-datail>

**Comment by zoechi** [comment]
@jcmordan 
GitHub issues are for bug reports and feature requests.
    For support questions please use other channels like the ones listed in [CONTRIBUTING - Got a Question or Problem?](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question)

**Comment by Parthchokshi** [comment]
Any update on this? It really is a headache. 

http://stackoverflow.com/questions/39925818/type-xyzcomponent-is-a-part-of-two-modules-module-gets-confused-what-route-to-t

**Comment by jredd** [comment]
Seems pretty silly to have to do this extra work to be able to use a component thats meant to be shared... Feels like an extra layer of configuration on top of all the other configuration thats needed...

**Comment by jcmordan** [comment]
as @brandonroberts saids, create a shared module like this:

```
import { NgModule }       from '@angular/core';
import { CommonModule }   from '@angular/common';
import { FormsModule }    from '@angular/forms';

import {SharedComponentA} from "../SharedComponentA";
import {SharedComponentB} from "../SharedComponentA";

@NgModule({
    imports: [
        CommonModule,
        FormsModule,

    ],
    declarations: [
      SharedComponentA,
      SharedCOmponentB

    ],
    providers: [
    ],
    exports: [
      SharedComponentA,
      SharedCOmponentB
    ]
})
export class SharedModule {}
```

then use the SharedModule like this..

```
import { NgModule }       from '@angular/core';
import { CommonModule }   from '@angular/common';
import { FormsModule }    from '@angular/forms';

import {SharedModule } from './SharedModule';

@NgModule({
    imports: [
        CommonModule,
        FormsModule,

        SharedModule

      //..
    ],
    declarations: [
     // ....
    ],
    providers: [
        // ....
    ]
})
export class PersonModule{}
```

**Comment by Yamilquery** [comment]
Domo arigato gozaimasu, from México! 👍

**Comment by mesopotamia** [comment]
I would have to agree that this is a big inconvenience. I have to create a module for something as trivial as one component.

**Comment by wsloth** [comment]
This is extremely inconvenient, as well as being plain ugly. You should really be able to inherit components from a parent module, because I think a shared module is extremely ugly.

You're going to end up with one huge shared module with all your components, which is one of the things I hate in more conventional platforms like .NET

I think you should be able to at least explicitly be able to share components between parent-child modules, in the sense of something like this:

```
@NgModule({
    imports: [
        CommonModule,
        FormsModule
    ],
    declarations: [
        Component1,
        Component2,
        // etc..
    ],
    expose: [ // Or really any word that would better describe what you're trying to do.
        Component1,
        Component2
    ]
})
export class AppModule{}
```

And then in your child module:
```
@NgModule({
    imports: [
        CommonModule,
        FormsModule
    ],
    inherit: [ // Inherit the exposed components from the parent module!
        Component1,
        Component2
    ]
})
export class SubModule{}
```

I'm sure there is a better way to do this. You just _don't want to have ten thousand components in a Shared module._

**Comment by EinarValholl** [comment]
@jcmordan
But how use different selectors from SharedComponentA and SharedComponentB in some template of PersonModule in different places?

**Comment by jcmordan** [comment]
@EinarValholl 
Just use the selector for each component on your html:

```
@Component({
    selector: 'component-a',
    template: '<div>IM COMPONNET A</div>'
})
class ComponentB{}

@Component({
    selector: 'component-b',
    template: '<div>IM COMPONNET B</div>'
})
class ComponentB{}

@Component({
    selector: 'edit-person',
    template: '<div>IM EDIT PERSON COMPONENT</div>
                   <div>
                          <component-a></component-a>
                          <component-b></component-b>
                     </div>'
})
class EditPersonComponent{}
```

**Comment by AyzPat** [comment]
how do i use my module as a reusable one in my main project in ng2?

**Comment by oeem1011** [comment]
I am also thinking about this. My Shared-Module has now nearly 30 components. Now i need to use one component in two different modules. So i need to import the whole Shared-Module twice in each component to only use one component? Is that not a big mistake against performance. What happens detailed during importing a module?

**Comment by oeem1011** [comment]
I have found an explanation in Angular Docs:

What if I import the same module twice?
That's not a problem. When three modules all import Module 'A', Angular evaluates Module 'A' once, the first time it encounters it, and does not do so again.

That's true at whatever level A appears in a hierarchy of imported modules. When Module 'B' imports Module 'A', Module 'C' imports 'B', and Module 'D' imports [C, B, A], then 'D' triggers the evaluation of 'C' which triggers the evaluation of 'B' which evaluates 'A'. When Angular gets to the 'B' and 'A' in 'D', they're already cached and ready to go.

Angular does not like modules with circular references so don't let Module 'A' import Module 'B' which imports Module 'A'.

**Comment by linusbrolin** [comment]
What if I have dozens of single-purpose components, that can be included by any number of other components (forms, subforms, views, etc)?
For example:
Consider an orderform, consisting of one component that has the customer information, one component that has a FormArray for each orderitem and then one component for each orderitem.
And for when the orderform is locked, I have the same amount of components for presenting just a view (cannot be changed by the user).

Do I have to create an NgModule for each and every one of those components?

If I create just one NgModule for all the orderitem related components, it means that loading a view component will also load FormModule/ReactiveFormModule for example..
That's completely unneccesary.

If I create a single NgModule for all of it, it will be quite bloated and that will result in longer load times, which is also bad. (I'm using http/2, so bundling into one single huge file is a bad thing for me).

So currently I see no other way than to make one NgModule for each and every single component/subcomponent/modal/etc file in the application (in order to not affect load times), which will force me to make hundreds of them. :/

**Comment by HFTSOL** [comment]
Why has there been no legitimate response to this?  Why is is still closed?  

This is a very big issue and needs to be fixed.  I see no point in touting the benefits of componentization, then immediately stuffing all those components into one module that has to be loaded each and every time, no matter what.  That's pathetic! 

Our adaptation of NG2 is being stalled by a couple of massive problems with NG2, and it would be nice for atleast one of them to be addressed in a meaningful way....  atleast let us know you are working to fix the problem.

**Comment by zoechi** [comment]
They built it that way for technical reasons. Don't expect thus to change. Rather embrace how modules work instead of beating a dead horse.

**Comment by Longfld** [comment]
brandonroberts:
"If you want to use a component across multiple modules, you'll need to create a "shared" module and add that component the shared module's exports. Then you add that shared module into your other modules imports"


there is performance issue in your way. it should, like lib for system, loaded when need and can be shared by all

**Comment by Fresst10** [comment]
"Shared" module work if componenet is componenet of a router, tlst it doesnt work.

**Comment by Longfld** [comment]
Fresst10, 
Yes, I am talking about component of a router, see my example (http://plnkr.co/edit/1fdAYP5PAbiHdJfTKgWo?p=preview) that I am going to use share module, could you please explain why there is not performance issues  inside brandonroberts' way?
And, of cause I would like these is no performance issue there.

**Comment by Fresst10** [comment]
I said about next situation.

"SharedModule"

import { ToolBarForm } from '../jsonSchemaForm/forms/toolbar/index';
import { NgModule }       from '@angular/core';
import { CommonModule }   from '@angular/common';
import { FormsModule }    from '@angular/forms';

@NgModule({
  imports: [
    CommonModule,
    FormsModule,
  ],
  declarations: [
    ToolBarForm
  ],
  exports: [
    ToolBarForm
  ]
})
export class SharedModule {}


"PersonalStaffDismissed" index.ts file

import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule, ApplicationRef } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PersonalStaffDismissed } from './staff';
import { SchemaFormModule } from "../../../../../srcs/schema-form/schema-form.module";
import { SharedModule } from '../../../../system_module/extendes_modules';

export const routes = [
  { path: '', component: PersonalStaffDismissed, pathMatch: 'full' } // (*)
];

@NgModule({
  declarations: [
    PersonalStaffDismissed,
  ],
  imports: [
    CommonModule,
    FormsModule,
    SchemaFormModule,
    SharedModule,
    RouterModule.forChild(routes),
  ]
})

In component "PersonalStaffDismissed" component "Toolbar" was nor declarated and it have error in browser.

For example, if i declareted "PersonalStaffDismissed" in "SharedModule" it was ok, because I used it in router (*).

**Comment by ammacdonald1** [comment]
Any updates? This doesn't really make sense. The workaround described breaks separation of concerns- rather than simply listing dependencies for your module and letting angular take care of injecting them you will end up just making a shared module that lists all the dependencies and import that into every module whether it needs 1 of the dependencies or 30.

**Comment by DzmitryShylovich** [comment]
Nobody forces u to create only one shared module with 30 deps. U can crate as many modules as u want

**Comment by lacma2011** [comment]
Who says we don't want to make multiple modules? The original problem is
that we can't share the same component across multiple modules. So now we
are forced to make one big shared module to make sure we don't run into
that problem.

On Thu, Jan 5, 2017 at 11:33 AM, Dzmitry Shylovich <notifications@github.com

**Comment by DzmitryShylovich** [comment]
@lacma2011 https://github.com/angular/angular/issues/10646#issuecomment-268873043 there's nothing to discuss here...

**Comment by Difinity** [comment]
@DzmitryShylovich 
It's immature syntactical idiocy to say you can create as many modules as you want when you know that whatever modules you create will ALL go into the same shared module.  The only real difference is one was is very cluttered and the other has significantly more code.  The problem remains!  

I'm not understanding why this is such a hard concept to understand.

@zoechi 
Saying this thread is beating a dead horse is the wrong expression....  It's more like complaining that my wife packs every bit of clothes, shoes, and accessories she owns every time she goes somewhere because she doesn't have the ability to just take the shit she's actually going to use. That's a better example.

**Comment by lacma2011** [comment]
How so? Is this not a legitimate feature request thread?

**Comment by DzmitryShylovich** [comment]
> whatever modules you create will ALL go into the same shared module.

absolutely not nessesary

**Comment by jcmordan** [comment]
Nobody said that you have to create just one shared module.
You can create as many shared modules as you need, grouping the component by functionaliy (for example).

Is more code? yes, but how much that you have to scare about?

Of course this behaivior could be better, there is but I thinks this make sense, if you think about optimization and permormance, remember that just one instance of a shared component is create it.

What do you prefere? less code or best performance?

**Comment by DzmitryShylovich** [comment]
If u want and have time u can re read this thread https://github.com/angular/angular/issues/10552#issuecomment-241575593
u will find reasons behind that desion

**Comment by Difinity** [comment]
@DzmitryShylovich
Read the comment linked.  he doesn't explain or justify the restriction.  He gives the thought behind shared module scoping and hierarchy, which is cool, but that's not what I've been talking about.

@jcmordan 
> What do you prefere? less code or best performance?
Both of course, but that's not the question here.  The problem here is that the current method is LESS performant than the desired method, not more.  The ability to load components and modules ONLY where they will be used, and not everywhere, is the desired behavior.  I haven't seen anyone say that components should be loaded multiple times if they are referenced multiple times, that's what caching is for.  

The seemingly undiscovered point is that we need not load and cache every module, component, pipe, directive, etc....  every frackin time the app loads.  I still haven't seen a cogent argument or explanation for this.

Also, I'm not continuing with this conversation hoping that NG2 will get fixed (that would be nice though).  I'm trying to find a reason for the madness.  So far, I have several major problems with the NG2 architecture and I am evaluating other frameworks to find which one will/might be better for our project for the long term.  So far, NG2 is still winning, but still doesn't help when the architecture is suffering.

**Comment by DzmitryShylovich** [comment]
> The problem here is that the current method is LESS performant than the desired method, not more.

prove it, then file a new issue with benchmarks.

**Comment by Difinity** [comment]
> 
>     The problem here is that the current method is LESS performant than the desired method, not more.
> 
> prove it, then file a new issue with benchmarks.

How would you suggest I do that?  Should I rewrite the framework to allow the desired behavior?

**Comment by Fresst10** [comment]
There is a problem with this module:
1 It is not possible to include two or more component
2 Do not work the trick with a share module
3 It is not possible to declare the component globally, to be visible in loadable
Due to the first two problems, the module is not possible to use angular2.
This is the important module for angular2.
If this is required to change the framework, I think it should be done only official.
Thx)

**Comment by kennygitchen** [comment]
Seems like community feedback is not going to be taken and we will have to load all the unnecessary code even if we only need one component in the SharedModule.

**Comment by zoechi** [comment]
@kennygitchen what's the problem with creating another shared module that contains only this one component?

**Comment by kennygitchen** [comment]
Lets say now I need this component in a newly created module, what should I do? Declare this "another-shared-module" into the root module? Consolidate this "another-shared-module" with another "shared-module"? (In large application consolidation will most likely create a domino effect) 

The main issue of the discussion of loading unnecessary component is not solved completely with this "another-shared-module" approach. It also adds complexity to the dependency management. 

I remember using RequireJS, where I only load the components that I need in each module. I don't have to load components that I don't need. 

And nowadays, people are using frameworks like Webpack where it is smart enough to create codesplits (bundle related JavaScript or resource files as one download) to help loading performance. With Webpack you are likely to load unnecessary components, but the framework is smart enough to minimise it (well to an extend, of course), you don't have to do it manually.

I am new to Angular, please forgive and correct me, if I am not making sense.

**Comment by mlc-mlapis** [comment]
In extreme cases you create one module per one component and load the module programmaticaly when you want or in combination with lazy loaded modules through router.

**Comment by zoechi** [comment]
@kennygitchen just add the modules where you want to use components, directives, or pipes from to `imports: [...]` of the module where you want to use them.

**Comment by JonnyBGod** [comment]
Extreme cases are... well, extreme. Not really good for the discussion.

I did not find any problem with the current system until now. I do not create any "shared" named module only feature named modules where I import needed dependencies. From time to time I do need to reorganize a bit, I mean change the point where I import a specific dependency or create a new module that makes sense to "modularise" a feature.

I actually feel that the current implementation points you to good practices and to make you reevaluate your architecture from time to time.

**Comment by kennygitchen** [comment]
@zoechi, I am not sure that is relevant, to add them into the module imports means it has already been added to the declarations in one of the parent or root modules.

@mlc-mlapis, an interesting idea, so it is a work-around for another work-around.

**Comment by mlc-mlapis** [comment]
@ennygitchen When you have a hundred products and each of them uses different combination of  attributes it is not a work-around but the basic principe.

**Comment by kennygitchen** [comment]
@mlc-mlapis, I think the principle is "declare it only once" as it is documented in the Angular documentation.

**Comment by zoechi** [comment]
@kennygitchen then just don't add it to parent or root module

**Comment by DaSchTour** [comment]
It's funny to see, that this issue will be there for ever. This very opinionated design decision will be a point to reconsider usage of other SPA-Frameworks. It's very interesting, that on the one hand angular allows a lot of configuration through decorators. On the other hand it adds a lot of hidden restrictions to this configuration and a lot of options to miss configure. So just at this point aurelia does a lot better job by removing all this configuration boilerplate. 

So somehow this discussion is a symptom of a unbalanced design. Adding a lot of freedom while extremely restricting it at the same time because of other more or less important (dependent of the project goals) design decisions.

**Comment by kennygitchen** [comment]
I am still study the framework but I have got the same feeling as @DaSchTour, and probably is why some of us are asking/challenging some of the design principles, trying to understand the reason behind these decisions.

Compare with frameworks like Aurelia, Angular 2 has more boilerplate. For example, if Angular used convention over configuration approach for things like css/html files (eg. app.component.ts automatically lookup app.component.html and app.component.css) with the config to specify them if needed. It probably removes most of the templateUrl/sytleUrls from @Component.

**Comment by chaiwa-berian** [comment]
Since I started learning Angular2, I have been confused a lil bit on the module concept and still struggling to understand the benefits other than the "folder benefits"! Can someone help me understand what qualifies a module to be a "shared module"? Does Angular have any way of identifying a module as "shared module"? Is every "child module" a shared module? 

I ask these questions, because I have failed to distinguish between what  I would call "no-shared module" and "shared module" from Angular specs other than "convention"! To me "feature modules" are setup the same way as "shared modules",the only difference usually is that I put fewer components/stuff in so called "shared modules" compared to what might be in "feature modules"! And importing a component from a "shared module" or from a "non-shared module or feature module" looks the same way, so it looks to me like it is just a question of "convention" and has nothing to do with "performance" and the like! Really I am so confused about the whole concept but I just have to hang on , being a man of faith! :smile:

**Comment by zoechi** [comment]
@chaiwa-berian a feature module and a shared module is the same thing. If you import a module it's a shared module ;-)

**Comment by hyspace** [comment]
The only way to solve this problem is have a well organized module structure, just like what angular itself did. However it is not possible in business logic. Some "pages" just share a similar UI component, for no reason and we can not group those component together like angular did because there are no relationship between them. If I started using "shared modules", finally it will end up with having a huge "shared module" that declares everything, to make sure nothing is declared twice.

The current solution will be have multiple "shared module" with reasonable size. This will avoid  importing all components when you only want to use one. However it is not convenient nor efficient.

**Comment by mlc-mlapis** [comment]
Multiple shared modules especially in combination with lazy loaded modules are important keys for good performance and easy maintenance. The main app module should be as optimized as possible.

**Comment by SBD580** [comment]
+1 

Separate module for each component is not practical. There should be a method to reuse components across different modules without creating for them a dedicated modules or some big shared module with all of them together (causing all of them to load when needing only one of them).

**Comment by mlc-mlapis** [comment]
So what is a real difference? Just these lines? What is so problematic on them?
```
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MyComponent } from './mycomponent'

@NgModule({
	declarations: [MyComponent],
	imports: [CommonModule]
})
export default class MyModule {}
```

**Comment by SBD580** [comment]
Technically speaking there is no difference and no problem. But from organizing aspect this is creating a lot of noise.
Imagine dozens of such components each requiring additional module file containing the exact same code.

I'm not saying this is impossible. I'm just saying there must be a better way.

**Comment by mlc-mlapis** [comment]
I fully understand. We have about 250 components = modules in one of our project. All loaded programmatically. To clear the organization part we have 1 file for the component + its module, so there is no difference from that point of view.

I think that ^^  those 7 lines that are repeated are not a real problem at all. I am using Angular 2 from alfa phase so I know what it was here before NgMopules. I am sure that they added more then we lost.

**Comment by SBD580** [comment]
Well I'm sure it could be worst of course, but I think we should think about how to make it better.

I think the real solution here, performance and organization wise, is the option to lazy load a component (and it deps). 
This way you won't be mind importing a huge module when only using several components and more importantly you won't be  loading + processing (parse/compile/storing etc) dozens of components for nothing in runtime.

**Comment by mlc-mlapis** [comment]
Yes, you are right. This is exactly what we are doing. We are programmatically lazy load only the module(s) = the component(s) that we need in one concrete moment. No more. Programatically because they using depend on business logic and not on routes but the same could be done via Angular router if really necessary. So still the only difference is ^^ 7 lines of code for defining the module structure for each module.

**Comment by mnn** [comment]
So for every single component which is used by two modules (e.g. two pages) I have to create a separate module? What an awful arbitrary restriction, it just requires us to copy-paste boilerplate, to create meaningless files.  :disappointed:

Combined with lazy loading the crash happens only when both pages (= modules) are accessed in same application run, which in my case was like a month. Very confusing...

**Comment by mlc-mlapis** [comment]
Once again `each component HAS TO be declared only in one NgModule`. It is not possible to have a component without including it to a module.

**Comment by mnn** [comment]
> each component HAS TO be declared only in **one** NgModule

I know that, that's the reason I wrote here... Why in just one? I don't see any logic behind having to create boiler-plate module for each component I want to be accessible in more modules (are we in Java world where boiler-plate is worshiped?). It feels like a random utterly unpractical restriction.

**Comment by mlc-mlapis** [comment]
There are certainly others but just the simplest one: `A component has a selector. You need an unique lookup for a component with the selector.`

Otherwise you will duplicate code, pile up problems with tree-shaking, need other rules which component factory should win when you have more identical, ...

**Comment by rob-balfre** [comment]
I wanted to figure out how to reuse a component between modules. Google brought me here. After reading this thread I just sighed. I'm not convinced by the shared module pattern described above. 

I'd love to know how the teams at Google using Angular on large apps structure their shared component code.

**Comment by rob-balfre** [comment]
FYI If you're like me and use angular-cli. When you refactor your code to use a shared module restart the cli server or you'll see console errors about the components not being registered.  It doesn't detect the changes otherwise. That had me stumped for a while!

**Comment by duannx** [comment]
Is there any update on this? One year latter since this issue was opened and i still can not find the answer.

**Comment by mlc-mlapis** [comment]
@duannx ... a component can be included in the declaration section only of one module - it is by design.

**Comment by duannx** [comment]
@mlc-mlapis Yeah. I khow it is by design but it is not the problem. The main question is how to share component between modules and the only answer is use share.module.ts. This way cause many annoying

**Comment by jcmordan** [comment]
@duannx I know is annoying, but usually an app  is build in a hirarchycal group of modules. With that in mind just put your shared component as top as you need, this way your component wil be avialable for all sub modules childrens of the module where you put your shared component.

**Comment by duannx** [comment]
@jcmordan Yeah. I think i should do it for my project. Thanks.

**Comment by Amarted** [comment]
This is angularJS javascrpt code, not typescript. but there is idea.
1. In ShopModule.js create mdoule:
angular.module('ShopModule', []);
2. In UserModule.js create mdoule:
angular.module('UserModule', []);
3. ShopService.js defines some service and we wont share without code duplicate. This is content of file:
// Some way to require ShopModule.js and UserModule.js 
...
// Add service to shop module
angular.module('ShopModule').service('ShopService', ShopService);
// Add service to user module
angular.module('UserModule').service('ShopService', ShopService);
// This is service
function ShopService {
    // some awesome code
}

**Comment by raysuelzer** [comment]
@jcmordan  -- I was wondering about this.  If I declare a component in my AppModule it should be available to any modules that need it that are in my imports?  That seems to be how it is working for me.

**Comment by zoechi** [comment]
@raysuelzer if you want to use components, directives, pipes from module B in module A, you have to import B (or any other module that exports B) in A. 
What you import in AppModule doesn't matter to module A except for services.

**Comment by rmcsharry** [comment]
@jcmordan In the case of lazy loaded modules I am not sure what you said is correct:

"With that in mind just put your shared component as top as you need, this way your component wil be avialable for all sub modules childrens of the module where you put your shared component."

If you have ModuleMain which has routes defined, and it lazy loads ModuleLazyA and ModuleLazyB and both of those need ThisComponent, putting ThisComponent into ModuleMain does not seem to work, even though hierarchically speaking it is above the two lazy modules.

I found that I needed to put ThisComponent into ThisSharedModule and import into BOTH of the lazy modules. Did I miss something?

**Comment by zoechi** [comment]
@rmcsharry  No, you need to import every module everwhere where you want to use components, directives, or pipes.
A module can export another module, this way you can import a set of modules with a single import. 
Besides that my first sentence still holds.

**Comment by rmcsharry** [comment]
@zoechi Thanks for the fast response. Are you saying I am correct or I am wrong? I thought that I could declare ThisComponent in ModuleMain and have it available to the two lazy loaded modules.

If ModuleLazyA and ModuleLazyB were NOT lazy loaded, then I presume that declaring ThisComponent inside ModuleMain (which would have to import A & B as NON lazy loaded) would then work, right?

It is only because A&B are lazily loaded that I CANNOT declare ThisComponent in ModuleMain, but must instead put that into a shared module for both lazy modules to import. This seems very counter-intuitive - since A&B are both lazy loaded by Main, why do they not get access to all components declared in Main?

**Comment by mlc-mlapis** [comment]
@rmcsharry ... generally the situation is simple ... you need to import (for example to `M`) always a module (for example `S`) when you want to use a component that is declared in that module (in `S`).

This the same principle both for eager or lazy module.

**Comment by rmcsharry** [comment]
@mic-mlapis It seems I am not doing a very good job of explaining a situation that appears to NOT follow these rules, yet should.

Here are the prerequisites for this problem:
1 ThisComponent is declared in Module S.
2 Module S declares two routes, A and B, each of which lazy load their own modules: ModuleLazyA, ModuleLazyB.

So in this situation, modules A and B are not importing Module S.
Module S is not importing either A or B, but is lazy loading them.
A and B are essentially children of S.

But neither A or B can see ThisComponent EVEN THOUGH it is declared in their parent module. They seem to just completely ignore it.

So I have to put ThisComponent into another module and then that module has to be imported by BOTH of the lazy modules.

Does that not seem odd to anyone else?

Why do I have to create another module just for ThisComponent...surely it makes more sense for the lazy loaded modules to have 'access' to whatever components are declared in their parent module?

If I am wrong and they CAN DO THAT, then clearly there is some extra plumbing code I need to add to make that happen.

If they CANNOT DO THAT then the statement by @jcmordan is incorrect when he said "your component will be available for all sub modules childrens of the module where you put your shared component."

**Comment by mlc-mlapis** [comment]
@rmcsharry ... you have to create a new module, even for this one component, we can call it SHARED, and this one will be imported to the eager module and both lazy loaded modules. The concept of modules is relative easy but what is SUPER IMPORTANT is the context in which components are declared, compiled, and used ...

And this is the reason why you need the new module ... and when lazy loaded modules will be loaded in future ... that new module will be already loaded before. Creating a new module is about 5 lines of code so it is nothing to care about.

**Comment by akariinc** [comment]
For someone who do not wants to use SharedModule like me, I got it working by setting separate tsconfig.json files for each builds. In tsconfig.anybuild.json file, set "exclude" options each entry file and module file and components / directives which is not necessary for the build.

**Comment by rmcsharry** [comment]
@mlc-mlapis Thank you for the reply but I believe you are still not understanding what I am saying, which is this:

1. You **don't need the shared module** if a parent is eager loading children. The parent can contain the component and both children will be able to use it.

2. You do need the shared module if a parent is lazy loading children. The children will not see the component if you declare it in the parent so in this scenario you **must use a shared module.**

So I would say it is something to care about - otherwise you are in danger of creating a huge shared module with maybe hundreds of components and then end up loading that on a lazy route just to get access to one single component. So how you structure shared code is quite important I think. I imagine most large Angular projects will require several shared modules, not just one, for this reason. And on routes which are not lazy loaded, they might not need a shared module at all - they can simply put the component in the parent.

**Comment by uloga** [comment]
I have the issue where I use angular material but I've built my app modular way for example I have AuthModule module with Auth components like login, register etc.  Now when I import angular material modules like for example MatButtonModule into my main app.module.ts I cannot use it in my AuthModule components since it's not recognizing the import from app.module.ts, so I have to import MatButtonModule into AuthModule to be able to use it in auth components!? Is this a bug or is it just how angular works!?  I'm using angular 5.

**Comment by mehmettkocc** [comment]
I have a shared module but I am getting the following. "ModuleX and ModuleX" 😄 

> Type PipeA is part of the declarations of 2 modules: ModuleX and ModuleX! Please consider moving PipeA to a higher module that imports ModuleX  and ModuleX. You can also create a new NgModule that exports and includes PipeA then import that NgModule in ModuleX and ModuleX .

**Comment by rjamesnw** [comment]
I'm doing research before I start using Angular CLI (first time dealing with Angular).  Based on this thread I'm deciding the best way to build my first Angular app is to create one module for every component or group of components that _perform a single function_, like:

1. All basic/primitive functionality controls have one module and one component per module.
    Example: 
        - 1 module and component for Avatar picture.
        - 1 module and component for an image file upload button.

2. Specialized "controls", such as a module that imports the avatar picture control (module) and the image upload control (module) together.

I believe thinking about modules as single "controls", not website "widgets" is the best way to modularize Angular apps (well, any web app really). That said, a widget of course will also be a module, so it's just a way of thinking (controls vs widgets [modules]).

**Comment by kennygitchen** [comment]
I don't think this thread is about what is the best practice, I think most of people here are not asking for what is the best practice to do this. Other than, this is quite a specific problem that people trying to get Angular to solve.

The issue here is really dependency management and probably, especially for people managing large applications with lazy loading components (loading components on the fly, so the app gets loaded in parts)

I have been working with a web application contains 60+ screens, and based on the configurations, each client gets different screens. Each screen implementation is complex enough to warrant its own module. So, we don't want to load the whole application when the user hit the url the first time. The whole app download will probably exceed 5MB+ in size. For us to be able to manage the 'dependencies' probably, it will be easier to just declare the components in the module and let the code/framework sort out the duplications. You can argue this is custom but as there are quite a few people coming in this thread to express their 'inputs' as 'this is not good enough' (well at least I am), is probably something worth look into.

In situation like this, declaring shared module specifically for the combination of 'screens' will be mad, therefore there is no choice but to put all the shared components into one single shared module. That defects the purpose of splitting the code. Of course there is always a solution to a problem, such as declare a shared module for each, for example client, but the point is 'we are giving our inputs to improve this framework'. 

This might be an old concept now as most of the web apps I have been working are just one JS file contains everything. Therefore, putting all shared component as one shared module isn't an issue at all. 

cheers, Kenny

**Comment by mlc-mlapis** [comment]
@kennygitchen ...

>  just one JS file contains everything / only one single shared module

Certainly not. Creating a module means adding 5 lines of code over the code for just a component, so it's nothing that should be a problem.

There is an idea to have a so-called `default` module (when Ivy will be ready) which should allow to lazily-load only a single component.

**Comment by rjamesnw** [comment]
@mlc-mlapis The default module wrapping sounds great, and makes sense.  I find myself wondering the same thing as I keep creating a module just to wrap a component for the purpose of sharing in other modules. Seems that should be implicit.

**Comment by phutaneVinayak** [comment]
> as @brandonroberts saids, create a shared module like this:
> 
> ```
> import { NgModule }       from '@angular/core';
> import { CommonModule }   from '@angular/common';
> import { FormsModule }    from '@angular/forms';
> 
> import {SharedComponentA} from "./SharedComponentA";
> import {SharedComponentB} from "./SharedComponentA";
> 
> @NgModule({
>     imports: [
>         CommonModule,
>         FormsModule,
> 
>     ],
>     declarations: [
>       SharedComponentA,
>       SharedComponentB
> 
>     ],
>     providers: [
>     ],
>     exports: [
>       SharedComponentA,
>       SharedComponentB
>     ]
> })
> export class SharedModule {}
> ```
> then use the SharedModule like this..
> 
> ```
> import { NgModule }       from '@angular/core';
> import { CommonModule }   from '@angular/common';
> import { FormsModule }    from '@angular/forms';
> 
> import {SharedModule } from './SharedModule';
> 
> @NgModule({
>     imports: [
>         CommonModule,
>         FormsModule,
> 
>         SharedModule
> 
>       //..
>     ],
>     declarations: [
>      // ....
>     ],
>     providers: [
>         // ....
>     ]
> })
> export class PersonModule{}
> ```

yes i try this way.

but angular giving error for shared components 

error: Template parse errors:
'trip-card' is not a known element:
1. If 'trip-card' is an Angular component, then verify that it is part of this module.

**Comment by Angelita31** [comment]
it works fine for me 
In my case: this is my component that is repeating ModalFilterOptionPage :
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Routes, RouterModule } from '@angular/router';

import { IonicModule } from '@ionic/angular';


import { ReactiveFormsModule } from '@angular/forms';

const routes: Routes = [
  {
    path: '',
/*     component: ModalFilterOptionPage remove it */
  }
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    IonicModule,
    RouterModule.forChild(routes)
  ]
/* declarations: [ModalFilterOptionPage] remove it*/
})
export class ModalFilterOptionPageModule {}

Then, only add it in your app module:
...
@NgModule({
  declarations: [
    AppComponent,
    ModalFilterOptionPage],
    entryComponents: [
    ModalFilterOptionPage
  ],

.....})
it works for me!

**Comment by aika-ssha** [comment]
Hey, guys! I have one problem. I have 8 lazy loaded modules and everything works perfect except 1 thing. When I logout from 1 module and login again with different role to other module it throws error, cause both of them use the same components from SharedModule

**Comment by michanagel** [comment]
I can only agree w/ @rmcsharry, he explained the problem multiple times in detail, specifically for lazy loaded modules.

question is also if I have to import that shared module again and again in lazy loaded modules (although the shared module was already loaded by the main app.module and should be accessible), will this bloat my lazy loaded chunks w/ redundant code ?
